flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/codegen/BaseAggregationCodeGenerator.scala

```diff
@@ -0,0 +1,776 @@
+import org.apache.flink.table.dataview.{StateListView, StateMapView}
+import org.apache.flink.table.functions.UserDefinedAggregateFunction
+import org.apache.flink.table.functions.aggfunctions.DistinctAccumulator
+import org.apache.flink.table.functions.utils.UserDefinedFunctionUtils
+import org.apache.flink.table.functions.utils.UserDefinedFunctionUtils.{getUserDefinedMethod, signatureToString}
+import org.apache.flink.table.runtime.aggregate.{GeneratedAggregations, GeneratedTableAggregations, SingleElementIterable}
+import org.apache.flink.table.utils.EncodingUtils
+import org.apache.flink.types.Row
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+/**
+  * A base code generator for generating [[GeneratedAggregations]] or
+  * [[GeneratedTableAggregations]].
+  *
+  * @param config                 configuration that determines runtime behavior
+  * @param nullableInput          input(s) can be null.
+  * @param inputTypeInfo          type information about the input of the Function
+  * @param constants              constant expressions that act like a second input in the
+  *                               parameter indices.
+  * @param name                   Class name of the function.
+  *                               Does not need to be unique but has to be a valid Java class
+  *                               identifier.
+  * @param physicalInputTypes     Physical input row types
+  * @param aggregates             All aggregate functions
+  * @param aggFields              Indexes of the input fields for all aggregate functions
+  * @param aggMapping             The mapping of aggregates to output fields
+  * @param distinctAccMapping     The mapping of the distinct accumulator index to the
+  *                               corresponding aggregates.
+  * @param isStateBackedDataViews a flag to indicate if distinct filter uses state backend.
+  * @param partialResults         A flag defining whether final or partial results (accumulators)
+  *                               are set
+  *                               to the output row.
+  * @param fwdMapping             The mapping of input fields to output fields
+  * @param mergeMapping           An optional mapping to specify the accumulators to merge. If not
+  *                               set, we
+  *                               assume that both rows have the accumulators at the same position.
+  * @param outputArity            The number of fields in the output row.
+  * @param needRetract            a flag to indicate if the aggregate needs the retract method
+  * @param needMerge              a flag to indicate if the aggregate needs the merge method
+  * @param needReset              a flag to indicate if the aggregate needs the resetAccumulator
+  *                               method
+  * @param accConfig              Data view specification for accumulators
+  */
+class BaseAggregationCodeGenerator(
```

Comment:
I like the idea of adding a base class of Aggregate code generator, But I think we can use the  `AggregationCodeGenerator` and only add `generateTableAggregations` is enough for table aggregate code gen. Because if we want to add a  base class of Aggregate code generator, we should consider how to deal with the` aggregates` vs `aggregate`, `distinctAccMapping` which table Aggregate never using.
So, I suggest if we really need add a base class of  Aggregate code generator, we can open a new PR.  What do you think?